portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.sps.data.Login; 
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    // returns empty string if user is logged in and the login url if user is not logged in  
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        String loginUrl=null, logoutUrl=null;
+        String userEmail=null;
+        Boolean loggedIn=userService.isUserLoggedIn();
+
+        // user not logged in, return login link
+        if (!loggedIn) {
+            loginUrl = userService.createLoginURL("/");
+        }
+
+        // user logged in, return logout link
+        else {
+            logoutUrl = userService.createLogoutURL("/");
+            userEmail = userService.getCurrentUser().getEmail();
+        }
+
+        Login loginDetails = new Login(loggedIn, loginUrl, logoutUrl, userEmail);
+
+        response.setContentType("application/json");
+        response.getWriter().println(convertToJsonUsingGson(loginDetails));
+        return;
+    }
+
+    // convert to JSON using GSON
+    private String convertToJsonUsingGson(Login loginDetails) {
+        Gson gson = new GsonBuilder().serializeNulls().disableHtmlEscaping().create(); 
```

Comment:
I don't know why it is necessary to use disableHtmlEscaping. Maybe you could set the response type to application/json, and that will fix the problem you are having?

Reply:
this is because '=' is changed to the unicode character '\u003d' when i did not use this disableHtmlEscaping method and that breaks the link. unfortunately, the login and logout url provided by the google API uses the '=' string.

more info is here: https://stackoverflow.com/questions/23363843/gson-disablehtmlescaping-why-gson-html-escapes-by-default-in-the-first-place/23364300

Reply:
thanks for the suggestion! disableHtmlEscaping is not used now